### ISSUE TITLE
Add a GDScript code formatter in the script editor

### ISSUE THREAD

**Issue by aki-cat** [godotengine/godot-proposals#1200]
**Describe the project you are working on:**
N/A.

**Describe the problem or limitation you are having in your project:**
Maintaining a uniform code style between files is time-consuming, and when working with multiple people, it becomes nearly impossible.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
[Original Issue](https://github.com/godotengine/godot/issues/7392)
_I'm just ressurrecting that issue because I have been watching it and didn't find any tracker for it anymore._

The desired functionality is to automatically reformat code in GDscript editor to some standard code style (like the one in the docs, for instance). Be it everytime the user saves the script or via a specific command, or maybe even both.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
A toggle in the editor menu would activate and deactivate this feature. You could configure it to reformat the file every time you save or you could configure a command that reformats the file to a hotkey.

This would greatly help with cleaning trailing whitespaces, enforcing line column limits, and generally keeping your code clean and uniform between developers.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It should be used fairly often by any coders with some experience. There is a third-party solution, however: [Godot GDScript toolkit](https://github.com/Scony/godot-gdscript-toolkit).

**Is there a reason why this should be core and not an add-on in the asset library?:**
GDScript is under constant changes and improvements. This kind of tool would work best if updated accordingly to the latest updates. Thus, being part of the core would be desirable.

**Comment by Calinou** [comment]
In the interest of making high-quality proposals, please [read the rules](https://github.com/godotengine/godot-proposals#readme) and fill in the issue template. This proposal will be reopened once this is done :slightly_smiling_face:

**Comment by OrenjiAkira** [comment]
Hello @Calinou !
I've updated the issue format accordingly. I apologize for not following right from the start, for some reason the template didn't load for me when I created the issue.

**Comment by shinspiegel** [comment]
This is a very good idea.
How should I proced to help this idea to become a reality?

**Comment by Calinou** [comment]
@shinspiegel Adding a code formatter is not a simple undertaking, so I'd recommend discussing this with @vnen if you have C++ programming experience.

**Comment by shinspiegel** [comment]
@Calinou, I've almost no experience with C++, but I'm willing to learn to improve the project.

**Comment by Calinou** [comment]
> @Calinou, I've almost no experience with C++, but I'm willing to learn to improve the project.

If you are looking at issues that are easy to fix for new C++ programmers, I would recommend looking at [issues labeled `junior job` on the main Godot repository](https://github.com/godotengine/godot/issues?q=is%3Aopen+is%3Aissue+label%3A%22junior+job%22).

**Comment by NathanLovato** [comment]
We're working on this! My teammate @Razoric480 started coding the formatter for Godot 4. The goal is to follow the official GDScript style guide and offer minimal options to customize the output (maybe line length and perhaps 1 vs 2 lines between functions, as Godot users have entire codebases using one or the other).

**Comment by NathanLovato** [comment]
After talking with the developers, we're gonna make a new proposal that explains the implementation details we're going for to make it easy for the GDScript maintainers to review and suggest a different approach.

**Comment by pycbouh** [comment]
Superseded by https://github.com/godotengine/godot-proposals/issues/3630 which proposes a specific implementation.
